### PR TITLE
DM/cli Sprint 2 stories

### PR DESCRIPTION
**Pull request by kopachevsky**
Implemented features:

- Cross-deployment referencing
-- Find references in yaml file
-- Replace reference occurences with actual output value
- Receive project name as input from CFT DM yaml  
- Input multiple DM templates at once 
- Implement Apply action
-- Implement Basic Create action
- Investigate profs and cons using gcloud or api with Basic Create action

### CODE REVIEW

**Comment by aaron-lane** [comment]
@dhobsd please review this for readability.

File: cli/deployment/deployment.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+func (d Deployment) ConfigFile() string {
+	return d.configFile
+}
+
+func NewDeployment(config Config, outputs map[string]map[string]string) *Deployment {
+	file, err := ioutil.TempFile("", config.Name)
+	defer file.Close()
+	if err != nil {
+		log.Fatal(err)
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: Here you only log the error. In some cases this error may be rather similar to an error from `file.Write()`, so adding some disambiguating context around the log call will be helpful to differentiate the two.

File: cli/deployment/deployment.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	return &Deployment{
+		config:     config,
+		configFile: file.Name(),
+	}
+}
+
+func replaceOutRefs(config Config, outputs map[string]map[string]string) []byte {
+	data, err := config.Yaml()
+	if err != nil {
+		panic(err)
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: Why not `log.Fatal` here?

File: cli/deployment/api_client_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		t.Error(err)
+	}
+	if "my-network-prod" != outputs["my-network-prod.name"] {
+		t.Errorf("Expected \"my-network-prod\" got \"%s\"", outputs["my-network-prod.name"])
+	}
+}
+
+// do actuall deployment creation call
+func testCreate(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (akopachevskyy-globallogic) [reply cli/deployment/api_client_test.go]: Yes, this one should not run during go test, I did use it to test real call with my local gcloud, maybe it make sense to remove it completely although I do  expect do use it couple times for debugging

File: cli/cmd/create.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		outputs[result.ID()] = result.Outputs
+	}
+}
+
+func loadConfigs(args []string) map[string]deployment.Config {
+	result := make(map[string]deployment.Config, len(args))
+	for _, path := range args {
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			log.Fatal("Error loading file", path)
```
[the remark this answers is not in the document]
Comment (akopachevskyy-globallogic) [reply cli/cmd/create.go]: well, technically it would be same output as log.Fatal("Error loading file", path) because Fatal will use default formatting for type, but I agree colon make more sense here

File: cli/deployment/api_client.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	outputs, err := GetOutputs(deployment.config.Name, deployment.config.Project)
+	if err != nil {
+		log.Print("Failed to get deployment Outputs", err)
+		return nil, err
+	}
+	deployment.Outputs = outputs
+	return deployment, nil
+}
+
+func ParseOutputs(data string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/api_client.go]: refactored to package level function

File: cli/deployment/api_client_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/api_client_test.go]: method removed

File: cli/deployment/api_client.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		log.Print("Failed to get deployment manifest", err)
+		return nil, err
+	}
+	return ParseOutputs(data)
+}
+
+func Create(deployment *Deployment) (*Deployment, error) {
+	_, err := execFunction("deployment-manager", "deployments", "create", deployment.config.Name, "--config",
+		deployment.ConfigFile(), "--project", deployment.config.Project)
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/api_client.go]: used args := []string option

File: cli/deployment/api_client_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	dep := Deployment{
+		configFile: "/1-prj/projects/google/CFT/fork/cloud-foundation-toolkit/dm/network.yaml",
+		config: Config{
+			Project: "gl-akopachevskyy-dm-seed",
+			Name:    "my-first-auto-deployment",
+		},
+	}
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/api_client_test.go]: as long it was real test, yes, but anyway I removed it

File: cli/deployment/config.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"log"
+	"regexp"
+	"strings"
+)
+
+var /* const */ pattern = regexp.MustCompile(`\$\(out\.(?P<token>[-.a-zA-Z0-9]+)\)`)
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config.go]: I would rather remove it

File: cli/deployment/order.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	}
+	g.outgoingNodes[name] = make(map[string]int)
+	g.incommingNodes[name] = 0
+	return true
+}
+
+func (g *DirectedGraph) AddNodes(names ...string) bool {
+	for _, name := range names {
+		if ok := g.AddNode(name); !ok {
+			return false
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/order.go]: I've refactored DirectedGraph struct and methods and made constructor method accept list of edges and nodes, no AddNode, RemoveNode, AddEdges anymore

File: cli/deployment/order_test.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
+package deployment
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestSort(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/order_test.go]: thanks. forgot to rename during refactoring

File: cli/deployment/order_test.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	if err != nil {
+		t.Error(err)
+	}
+	expected := []Config{configA, configB}
+
+	if !reflect.DeepEqual(actual, expected) {
+		t.Errorf("Expected %v, got %v", expected, actual)
+	}
+
+}
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/order_test.go]: I did API private, but I agree I need to add more edge cases here, thanks for pointing me to table tests

File: cli/deployment/config_test.go [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+	if len(config.Imports) != 1 {
+		t.Errorf("Expected to have 1 import")
+	}
+	if len(config.Resources) != 2 {
+		t.Errorf("Expected to have 2 resources")
+	}
+}
+
+func TestFindAllOutRefs(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config_test.go]: refactored to table-driven test, thanks

File: cli/deployment/config_test.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+	if configA.findAllDependencies(configs) != nil {
+		t.Errorf("ConfigA should not have any deps")
+	}
+
+	if !reflect.DeepEqual(configB.findAllDependencies(configs), []Config{configA}) {
+		t.Errorf("ConfigB should have ConfigA as dep")
+	}
+}
+
+func TestYaml(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config_test.go]: that's means I need to rename function to YAML()? looks quite ugly to me

File: cli/deployment/deployment.go (hunk 2 above)
Comment (kopachevsky) [reply cli/deployment/deployment.go]: atavism left from debug

File: cli/deployment/config.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func (c Config) findAllOutRefs() []string {
+	matches := pattern.FindAllStringSubmatch(c.data, -1)
+	result := make([]string, len(matches))
+	for i, match := range matches {
+		result[i] = match[1]
+	}
+	if len(result) == 0 {
+		return nil
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config.go]: is it rule true not just for arrays but for struct objects as well? I mean if we'll return struct instance anyway with lets fields initialized with default values, isn't it error prone?

**Comment by kopachevsky** [comment]
> Hi there! Thanks for attaching me to this review.
> 
> There are a lot of comments here, but I think the majority of them follow similar form, so once you're in the swing of things, it may be easy to address all of the comments of a particular class. I've attempted to attach links to relevant documentation about code review comments and common idioms where appropriate, so you can also incorporate that into future work. Most of this review is not for correctness, though I have pointed out a few cases in which some APIs might be improved or better tested. It would likely be beneficial to add another reviewer for correctness as well.
> 
> Please let me know if you have any questions!

@dhobsd Devon, thanks for time your spent on review, I agreed and applied changes to most of the comments, only 3 left with my comments, please reply if you'll have spare moment.

File: cli/deployment/config.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (dhobsd) [reply cli/deployment/config.go]: To be clear, this function returns a slice rather than an array type.

This code review suggestion is incorrect on my part, probably due to some mental gymnastics I didn't do properly when looking at the callers of this function. In fact, the [Code Review Comments wiki on empty slices](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices) explicitly states that it's idiomatic to prefer a `nil` slice over an empty slice in most cases. So apologies for the confusion here.

Instead, the callsite in `findAllDependencies` that tests `refs != nil` should instead be testing `len(refs) > 0`. That would obviate the need for the `nil` return. And this function body could look like:

```go
	var result []string
	matches := pattern.FindAllStringSubmatch(c.data, -1)
	for i, match := range matches {
		result = append(result, match[1])
	}
	return result
```

The function _may_ still return `nil` in the case that `len(matches) == 0`, but that's fine. Making the distinction between the nil slice and the empty slice in the `findAllDependencies` test results in additional logic here that ultimately isn't needed.

To answer the question on structs, this comment really only applies to pointer types, which include pointers, slices, maps, channels, functions, and interfaces. In Go, any type is implicitly initialized to its zero value. For pointer types, this is `nil`. Dave Cheney has a [good blog post on the zero value and making it useful](https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful).

The reason I look out for this sort of thing is that semantics of the zero value differ for different types, and dereferencing `nil` is a runtime error -- so that's potentially problematic for pointer types.

So why not for slices? Slices are often manipulated and accessed through builtin APIs. For example, ranging over a `nil` slice is the same as ranging over a 0-length slice: the range returns nothing. The `len` and `cap` of a `nil` slice will return 0. `Append`ing to a nil slice will magically create storage. This is why `nil` is fine in most cases for slices.

Maps, functions, and interfaces are a bit different. Maps don't have builtin APIs for most accesses; inserting into or reading from a `nil` map is a runtime error. Function-typed objects with `nil` values cannot be called; this is a runtime error. `Nil` interfaces are especially confusing; there's even a [FAQ on why a nil error value is not equal to nil](https://golang.org/doc/faq#nil_error) (though the text there applies to interfaces in general, it's often encountered when dealing with `error`, which is probably the most commonly used interface type).

This subtlety also extends to pointer types. Methods with pointer receivers (e.g. `func (t *T) Foo() error`) may expect that pointer to be non-`nil`. However, you may still call a method on a `nil`-valued type (e.g. `var t *T; t.Foo()`). If T is (say) a struct, and `(*T).Foo` accesses fields in that struct, this is also a runtime error.

So that's where this commentary is coming from, and things to watch out for when dealing with pointer types in Go.

File: cli/deployment/config_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (dhobsd) [reply cli/deployment/config_test.go]: Yes, this is convention in Go. C.f. `net.IP`, `http.ServeTLS`, etc.

Although it's not an initialism, this convention is adopted for `ID` in most cases as well.

File: cli/deployment/utils_test.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	expected = "/base/templates/script.py"
+	actual = AbsolutePath("/base/folder/config.yaml", "../templates/script.py")
+	if actual != expected {
+		t.Errorf("Expected %s, actual %s", expected, actual)
+	}
+}
+
+func GetTestData(folder string, name string, t *testing.T) string {
+	path := filepath.Join("../testdata", folder, name)
+	buff, err := ioutil.ReadFile(path)
```
[the remark this answers is not in the document]
Comment (dhobsd) [reply cli/deployment/utils_test.go]: `buf` is a more common name than `buff`. It's admittedly a nit, but it comes from a [convention for preference of short variable names](https://github.com/golang/go/wiki/CodeReviewComments#variable-names).

[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/utils_test.go]: got it, thanks

File: cli/deployment/config.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config.go]: thanks for such deep explanation!

File: cli/cmd/create.go [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+	"github.com/spf13/cobra"
+
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/deployment"
+)
+
+var projectFlag string
+
+func init() {
+	createCmd.PersistentFlags().StringVarP(&projectFlag, "project", "p", "", "project name")
```
Comment (ocsig) [inline cli/cmd/create.go]: project name >> ProjectID

File: cli/README.md [hunk 17]
```diff
@@ -4,7 +4,7 @@
 
 ### Requirements 
 
- * go 1.12
+ * go 1.12 and above 
 
 ### Build and Run
```
Comment (ocsig) [inline cli/README.md]: After 'make build' customer should be able to call  'cft' directly without the 'bin/' prefix ( or any other prefix) can we add it to the global path?

Comment (ocsig) [inline cli/README.md]: Missing [parameters]: 
  cft [flags]
  cft [command] [flags]

Comment (kopachevsky) [reply cli/README.md]: I thought end user have to upload cli executable manually, not build it with GO. Even in case if we'll ask end user to build it he'll should rather do this with 

`go get https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/tree/master/cli` command, which will place built binary to $GOPATH/bin

Comment (ocsig) [reply cli/README.md]: Thats fine, but that needs to be documented, how to use the CLI as a total new user.

Comment (kopachevsky) [reply cli/README.md]: Agreed, I would propose to do documentation as last task, once all functionality will be implemented

**Comment by dhobsd** [comment]
Thanks for taking care of these! I have a few more comments that I've been working on as I've had time -- mostly to do with (heh) comments, but this is looking much better! I'm hoping to have one more round of review ready today or tomorrow. Just wanted to add a quick update here so that silence isn't taken as ignoring the PR or implicit approval.

File: cli/deployment/config.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	return c.Project + "." + c.Name
+}
+
+func (c Config) String() string {
+	return c.FullName()
+}
+
+// YAML function converts Config object to yaml
+// overrides all relative paths for imports to absolute form,
+// removes all custom elements gcloud deployment manager not aware of (name, project, description)
```
Comment (dhobsd) [inline cli/deployment/config.go]: "The YAML function marshals a Config object to YAML. It sets all relative import paths to absolute paths and removes all custom elements that the gcloud deployment manager is not aware of, including name, project, and description."

File: cli/deployment/deployment.go [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"strings"
+)
+
+// Deployment object represent real GCP Deployment entity that already created or have to be created
+// configFile field point to yaml file, generated from Config object with all cross-deployment references
+// overwritten with actual values
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: "The Deployment type represents a real GCP Deployment entity that is either already created, or yet-to-be created."

The bit about configFile should be attached to the `configFile` member, fixing YAML casing and as a complete sentence.

File: cli/deployment/deployment.go [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+type Deployment struct {
+	Outputs    map[string]string
+	config     Config
+	configFile string
+}
+
+// NewDeployment creates Deployment object and override all out refs, this means all
+// deployments it depends to should exists in GCP project
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: "NewDeployment creates a new Deployment object, overriding all outward references. In effect, this means all deployment dependencies must exist in the GCP project."

File: cli/deployment/deployment.go [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+		log.Fatalf("Error wirte to file: %s, error: %v", file.Name(), err)
+	}
+
+	return &Deployment{
+		config:     config,
+		configFile: file.Name(),
+	}
+}
+
+// String representation of Deployment instance
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: Doc comments are often elided for implementations of common interfaces like `fmt.Stringer` (which this does). But if you keep the comment, it should say something to the effect of "String implements fmt.Stringer for the Deployment type."

File: cli/deployment/deployment.go [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+func (d Deployment) String() string {
+	return fmt.Sprintf("Deployment[name=%s, config=%s]", d.config.FullName(), d.configFile)
+}
+
+// same as deployment.Config.FullName(), can be used in map[string]Deployment as a key
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: This should be a full sentence, and should duplicate the text for `config.FullName`. The generated documentation won't be cross-referenced, so it would be helpful for someone looking here to see the full behavior description rather than needing to go find the `FullName` method on the `config` type.

File: cli/deployment/doc.go [hunk 23]
```diff
@@ -0,0 +1,7 @@
+/*
+Package Deployment implements GCP Deployment related functionality:
+sort deployment configs in dependency order,
+execute API calls to create, update, delete, describe deployment using gcloud tool
+ */
```
Comment (dhobsd) [inline cli/deployment/doc.go]: "The deployment packages implements functionality for deploying projects to GCP. This includes sorting deployment configurations in dependency order, and executing GCP API calls through the gcloud tool."

File: cli/deployment/order.go [hunk 24]
```diff
@@ -0,0 +1,137 @@
+package deployment
+
+import (
+	"errors"
+	"fmt"
+	"log"
+)
+// Order function receives map of configs with Config.FullName() string as a key,
+// find dependencies between them, and order them topologically using directed graph.
+// Returns array of ordered config's objects and error if configs have circle dependencies
```
Comment (dhobsd) [inline cli/deployment/order.go]: "The order function receives a map of Configs keyed by each configuration's FullName, resolves dependencies between them, and sorts them topologically. It returns the sorted array of Configs, or an error if dependencies are cyclic."

File: cli/deployment/order.go [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+	res := make([]Config, size)
+	for i, name := range sorted {
+		res[i] = configs[name]
+	}
+	return res, nil
+}
+
+
+// directedGraph struct used to build graph of cross-config depencencies
+// and do topological sort to define ordering of deployment creation
```
Comment (dhobsd) [inline cli/deployment/order.go]: Note I'm not commenting on the grammatical structure of these comments because they are attached to non-exported types and functions, but you might like to address these as well to assist with readability for folks developing the package itself.

File: cli/deployment/api_client.go [hunk 26]
```diff
@@ -0,0 +1,118 @@
+package deployment
+
+import (
+	"bytes"
+	"fmt"
+	"gopkg.in/yaml.v2"
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/api_client.go]: by own section you mean empty line before "gopkg.in/yaml.v2"?

File: cli/deployment/config_test.go [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+
+	if len(config.Imports) != 1 {
+		t.Errorf("expected 1 import, got: %d", len(config.Imports))
+	}
+	if len(config.Resources) != 2 {
+		t.Errorf("want 2, got: %d", len(config.Resources))
+	}
+}
+
+var outRefTests = []struct {
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config_test.go]: make sense, I was thinking about this as well

File: cli/deployment/api_client.go [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// function exposed to variable in order to mock it inside api_client_test.go
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/api_client.go]: so it's good style to mention function name in commend?

File: cli/deployment/config.go [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+
+	"gopkg.in/yaml.v2"
+)
+
+// patern to parse $(project.deployment.resource.name) and $(deployment.resource.name)
+var pattern = regexp.MustCompile(`\$\(out\.(?P<token>[-.a-zA-Z0-9]+)\)`)
+
+// Config struct keep config data parsed from passed config.yaml
+type Config struct {
+	Name        string
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/deployment/config.go]: I agree doc is important!

File: cli/deployment/api_client.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (dhobsd) [reply cli/deployment/api_client.go]: Yep!

File: cli/deployment/api_client.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (dhobsd) [reply cli/deployment/api_client.go]: It is! The [Effective Go section on "Commentary"](https://golang.org/doc/effective_go.html#commentary) addresses best practices for documentation and doc comments. You're likely to see suggestions that follow these recommendations in code reviews.

**Comment by dhobsd** [comment]
This is looking good to me from a readability and base correctness perspective; please feel free to complete the review for functionality / semantics! (I'm happy to give another once-over when that's done.)

File: cli/cmd/doc.go [hunk 30]
```diff
@@ -0,0 +1,5 @@
+/*
+Package cmd contains github.com/spf13/cobra commands implementing CFT CLI logic
+ */
```
[the remark this answers is not in the document]
Comment (kopachevsky) [reply cli/cmd/doc.go]: Updated with punctuation mostly

**Comment by kopachevsky** [comment]
> This is looking good to me from a readability and base correctness perspective; please feel free to complete the review for functionality / semantics! (I'm happy to give another once-over when that's done.)

Thanks, please do another review, p.s. I'll do final round of all comments and revisit documentation as the very last step of development again, as long I did lots of refactor already (see https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/114)

**Comment by kopachevsky** [comment]
@dhobsd can you take another look and approve if it's fine

**Comment by dhobsd** [comment]
I don't appear to have the capability to approve the PR.

**Comment by morgante** [comment]
Sounds like @dhobsd approves the code.

@ocsig Can you give the final functional approval?
